core/jvm/src/main/scala/fs2/interop/flow/StreamUnicastPublisher.scala

```diff
@@ -0,0 +1,103 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package interop
+package flow
+
+import cats.effect.kernel.{Async, Deferred, DeferredSink, DeferredSource, Resource}
+import cats.effect.std.Dispatcher
+import cats.syntax.all._
+
+import java.util.concurrent.Flow.{Publisher, Subscriber}
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.util.control.NoStackTrace
+
+/** Implementation of a [[Publisher]].
+  *
+  * This is used to publish elements from a [[Stream]] to a downstream reactive-streams system.
+  *
+  * @see [[https://github.com/reactive-streams/reactive-streams-jvm#1-publisher-code]]
+  */
+private[flow] final class StreamUnicastPublisher[F[_], A] private (
+    stream: Stream[F, A],
+    startDispatcher: Dispatcher[F],
+    requestDispatcher: Dispatcher[F],
+    completedSignal: DeferredSink[F, Unit],
+    usedCheck: AtomicBoolean
+)(implicit
+    F: Async[F]
+) extends Publisher[A] {
+  def subscribe(subscriber: Subscriber[_ >: A]): Unit = {
+    nonNull(subscriber)
+    startDispatcher.unsafeRunAndForget {
+      F.delay(usedCheck.getAndSet(true)).flatMap {
+        case false =>
+          StreamSubscription(
+            subscriber,
+            stream,
+            startDispatcher,
+            requestDispatcher,
+            completedSignal
+          ).flatMap { subscription =>
+            F.delay {
+              subscriber.onSubscribe(subscription)
+              subscription.unsafeStart()
+            }
+          }
+
+        case true =>
+          // This Publisher can only be used once.
+          F.delay(
+            subscriber.onError(
+              StreamUnicastPublisher.StreamUnicastPublisherAlreadyUsedException
+            )
+          )
+      }
+    }
+  }
+}
+
+private[flow] object StreamUnicastPublisher {
+  def apply[F[_], A](
+      stream: Stream[F, A]
+  )(implicit F: Async[F]): Resource[F, (StreamUnicastPublisher[F, A], DeferredSource[F, Unit])] =
+    (
+      Dispatcher.sequential[F],
+      Dispatcher.sequential[F],
+      Resource.eval(F.delay(new AtomicBoolean(false))),
+      Resource.eval(Deferred[F, Unit])
+    ).mapN { case (startDispatcher, requestDispatcher, usedCheck, completedSignal) =>
+      new StreamUnicastPublisher(
+        stream,
+        startDispatcher,
+        requestDispatcher,
+        completedSignal,
+        usedCheck
+      ) -> completedSignal
+    }
+
+  private object StreamUnicastPublisherAlreadyUsedException
```

Comment:
Maybe this exception should be moved to the `package` and made public?

Reply:
According to the JavaDoc you should be throwing an `IllegalStateException`.
https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.Publisher.html#subscribe-java.util.concurrent.Flow.Subscriber-

Reply:
Not throwing but sending :)

But yes, I didn't realize it should be an `IllegalStateException`, will fix that.
Also, that means there is no need to make this one public.